truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java

```diff
@@ -0,0 +1,716 @@
+    }
+
+    private <T extends EventListener> EventBinding<T> attachListener(AbstractInstrumenter instrumenter, SourceSectionFilter filter, T listener) {
+        return addBinding(new EventBinding<>(instrumenter, filter, listener));
+    }
+
+    private static boolean isInstrumented(Node node, EventBinding<?> binding, SourceSection section) {
+        return binding.getInstrumenter().isInstrumentable(node) && binding.getFilter().isInstrumented(section);
+    }
+
+    private static boolean isInstrumentedRoot(RootNode node, EventBinding<?> binding, SourceSection section) {
+        return binding.getInstrumenter().isInstrumentable(node) && binding.getFilter().isInstrumentedRoot(section);
+    }
+
+    private static boolean isInstrumentedLeaf(Node node, EventBinding<?> binding, SourceSection section) {
+        if (binding.getFilter().isInstrumentedNode(section)) {
+            assert isInstrumented(node, binding, section);
+            return true;
+        }
+        return false;
+    }
+
+    private static void trace(String message, Object... args) {
+        PrintStream out = System.out;
+        out.printf(message, args);
+    }
+
+    private static void visitRoot(final RootNode root, final AbstractNodeVisitor visitor) {
+        if (TRACE) {
+            trace("Visit root %s wrappers for %s%n", visitor, root.toString());
+        }
+
+        if (visitor.shouldVisit(root)) {
+            // found a filter that matched
+            root.atomic(new Runnable() {
+                public void run() {
+                    root.accept(visitor);
+                }
+            });
+        }
+        if (TRACE) {
+            trace("Visited root %s wrappers for %s%n", visitor, root.toString());
+        }
+    }
+
+    static void removeWrapper(ProbeNode node) {
+        if (TRACE) {
+            trace("Remove wrapper for %s%n", node.getContext().getInstrumentedSourceSection());
+        }
+        WrapperNode wrapperNode = node.findWrapper();
+        ((Node) wrapperNode).replace(wrapperNode.getDelegateNode());
+    }
+
+    private static void invalidateWrapper(Node node) {
+        Node parent = node.getParent();
+        if (!(parent instanceof WrapperNode)) {
+            // not yet wrapped
+            return;
+        }
+        invalidateWrapperImpl((WrapperNode) parent, node);
+    }
+
+    private static void invalidateWrapperImpl(WrapperNode parent, Node node) {
+        ProbeNode probeNode = parent.getProbeNode();
+        if (TRACE) {
+            SourceSection section = probeNode.getContext().getInstrumentedSourceSection();
+            trace("Invalidate wrapper for %s, section %s tags %s%n", node, section, Arrays.toString(section.getTags()));
+        }
+        if (probeNode != null) {
+            probeNode.invalidate();
+        }
+    }
+
+    static Instrumentable getInstrumentable(Node node) {
+        Instrumentable instrumentable = node.getClass().getAnnotation(Instrumentable.class);
```

Comment:
This seems to get only the annotations of the current class. Is that intentional?
I tried to annotate an `Invokable` node class with `InstrumentationTag.ROOT` and I was surprised that this is not recognized in the `Method` and `Primitive` subclasses.

This is especially surprising since this behavior seems to be different from the TruffleDSL annotations.